index.rst

```diff
@@ -48,14 +48,177 @@
+This would lead to a situation where a workflow can complete but the registry it out of date for an indeterminate period of time and would delay submission of workflows that depend on the results.
+Were that to happen though, it would be indicative that letting each workflow attempt the sync up directly at the end would be risky.
+Using a queue also completely removes registry credentials from workflow execution since the queued updates would be running in a completely different environment.
+
+The synchronization must check that a corresponding dataset was written to the object store since it is possible for a workflow to partially complete and synchronization should (optionally?) happen even on failure since that can allow resubmission of the workflow with different configurations and also allow investigation of the intermediate products.
+
+The limited registry approach still requires that the butler is interacting with an object store and would require the use of signed URLs if the datasets were being written to their final location in the shared datastore.
+If the signing client becomes a bottleneck an alternative could be to use a temporary bucket specifically for that workflow.
+
+The synchronization job would therefore also need to move the files from the temporary location to the final location.
+
+
+Summary
+=======
+
+Whilst the job-level pooling and externalized approaches will help with registry contention, by far the safest approach, and in the long term the simplest, is to adopt the limited read-only registry.
+This removes any worries about merging the outputs of multiple jobs and lets us leverage the knowledge already known to the graph builder.
+
+Task Breakdown
+--------------
+
+The initial prototype implementation will not include URL signing since all Data Preview 0.2 workflow executions will be run by staff.
+
+An initial break down of work would then be:
+
+* Add facility to pipetask command line to create a minimalist SQLite registry from the constructed quantum graph.
+  Initially this registry could ignore the datastore table completely.
+* Create a new ``Datastore`` subclass that no longer queries registry on ``Datastore.get()`` but instead assumes that the active datastore configuration (read from the user-supplied Butler configuration) would give the correct answer (something that can not be relied on in general but can be relied on in this limited context) for file template and formatter class.
+  Also modify ``Datastore.put()`` such that the registry is never written to (this could be implemented as a new limited datastore registry manager class).
+* Either change ``pipe_base`` Butler interface to use an entirely new implementation of ``ButlerQuantumContext`` that ignores registry and uses the new ``Datastore`` class directly, or else implement a new ``Registry`` subclass that for write operations instead compares the supplied values with the expected values to ensure self-consistency.
```

Comment:
Or a new `Butler` class.

Or maybe just logic that `Butler.put` writes to datastores if the dataset is already in registry but not in one of its datastores; that is pretty much what we're asking for here, and it seems like not-unreasonable behavior in general.

Reply:
A butler configuration option in the YAML that says "always ask registry to see if we have dataset and if we have directly do the datastore.put" works. Your comment about "only if we have not stored the file yet" depends on datastore registry entries not being filled in by graph builder. See comment above about that.

Reply:
Yup, I would want to just do `Registry.insertDatasets` in QG gen and never populate the `Datastore` records in the datastore until the files actually exist.  Getting the `dataset_id` values earlier is what really helps, and the system is at least theoretically designed to be able to handle "dataset exists but not in this datastore" already (though we hadn't really thought about that implying "exists but not necessarily in any datastore", as we're discussing now).